submariner/IPV6-datapath.md

```diff
@@ -0,0 +1,190 @@
+It shares these with other clusters connected to the Broker.
+It also ensures that it is the sole Endpoint for its cluster.
+The Endpoint resource fields should include IP addresses according to the cluster’s networking configuration.
+For example for a dual-stack cluster HealthCheckIP, PrivateIP, PublicIP and Subnets should consist of both IPv4 and IPV6 addresses.  
+
+The Route Agent running in the cluster learns about the local Endpoint and remote Endpoints.
+It sets up the necessary V4,V6 infrastructure to route cross-cluster traffic from all nodes to the active Gateway Engine node.
+
+The active Gateway Engine establishes a watch on the Broker to learn about Endpoint and Cluster resources from other clusters.
+Once two clusters are aware of each other’s Endpoints, they can establish secure tunnels.
+These tunnels are based on the remote and local Endpoint details, allowing traffic to be routed.
+A tunnel should be created only if the local Endpoint's networking type matches the remote Endpoint's IP family.
+
+The next diagram illustrates Submariner’s datapath architecture for kube-proxy based CNIs:
+![non-ovnk-architecture](./images/dual-stack-arch-for-non-ovnk.png)
+
+With the proposed architecture, Submariner needs to establish both V4 and V6 intra-cluster egress routing to the GW node in the case of dual-stack.
+
+Pod IPV4 egress packets for CNI != OVNK and cable-driver=libreswan will be:
+![non-ovnk-ipv4-egress](./images/ipv4-non-ovnk-egress-packets.png)
+
+And Pod IPV6 egress packets for the same configuration will be:
+![non-ovnk-ipv6-egress](./images/ipv6-non-ovnk-egress-packets.png)
+
+For IPV4 VxLAN encapsulation we use 242.x.x.x CIDR range, a similar IPV6 CIDR should be used for IPV6 VxLAN encapsulation.
+
+**Note**: In future, we may optimize this architecture for a dual-stack case.
+For example by using only the intra-cluster V4 VxLAN to route V4 and V6 traffic to the GW.
+
+## Datapath breakdown
+
+### Gateway
+
+To support IPV6 the gateway should:
+
+* discover publicIP, privateIP, healthcheckIP and cluster’s subnets for each IP family.  
+  * Note: The gateway should address corner cases related to this change.
+  For example, in a dual-stack environment, only the V4 public IP address might be successfully resolved.
+* run NAT Discovery per IP family in remote Endpoint.
+* Advertise IP details in the local Endpoint based on the cluster's networking type.
+For example, in a dual-stack cluster, both V4 Public IP and V6 Public IP should be advertised in the Endpoint.
+* continue advertising a **single** Endpoint. in case of a dual-stack cluster, fields should consist of both V4 and V6 addresses separated.
+
+**Note**: Three new fields (HealthCheckIPs, PrivateIPs, PublicIPs) should be added to the Endpoint resource to support dual-stack.
+ These fields will store multiple IP addresses for dual-stack scenarios.
+
+* create inter-cluster tunnel only if local endpoint networking type matches remote endpoint ipfamily.
```

Comment:
Lets say we have two dual-stack clusters A and B. If the clusters are able to successfully connect over "IPv4" but fail to connect over "IPv6", what is the impact on the connection status (f.e., the user facing commands like subctl show xxx). Please add some notes about this scenario.

Reply:
Done.